### ISSUE TITLE
Export environment `OUT_DIR` to proc-macro.

### ISSUE THREAD

**Issue by Neutron3529** [rust-lang/cargo#14035]
### Problem

Currently, crates-io will yield such error:
```
error: failed to verify package tarball

Caused by:
  Source directory was modified by build.rs during cargo publish. Build scripts should not modify anything outside of OUT_DIR.
  Added: /me/rmin/target/package/rmin-examples-0.0.1/aaa.rmin.Rust.Functions.R
```

But since I only have a `proc-macro` and no `build.rs`, the `OUT_DIR` is not set. And since I need some of their output (generating R wrappers for further roxygen2 doc generation and further package loading), the output is necessary.

Is there some configuration that could make a `OUT_DIR` visible for `proc-macro`s?

### Proposed Solution

Currently, for build.rs, it could write output to `OUT_DIR`, but for `proc-macro`-only crate, such option does not exists.

Another solution is that adding a [cache] section in Cargo.toml, tell Cargo what file could be written. With this configure, we may have the ability to debug proc-macros by tell them writting down all the code they generates.

**Comment by epage** [comment]
See also [Internals discussion](https://internals.rust-lang.org/t/proc-macro-output-dir/21001)

**Comment by Ddystopia** [comment]
This would be helpful for image preprocessing. You can't just feed a huge image into the token stream, better would be to write them to the file system and use `include_bytes!`.

**Comment by epage** [comment]
`OUT_DIR` likely should be considered writeable only by the build script due to parallel builds and shared caching. 

Is this just for a scratchpad or does this create final artifacts? If scratchpad, https://github.com/rust-lang/cargo/issues/16427 may be relevant.

**Comment by Ddystopia** [comment]
For parallel builds you can just create a new subdir for each contender, isn't it? Of course it is expected that they won't `../` out of them, this behavior should be marked as unspecified. This way each process can have final artifacts while not conflicting.

**Comment by epage** [comment]
`OUT_DIR` is created when build scripts are present for them to write into and others to read. Setting another value for proc-macros would be a breaking change. If we conditioned it on the absence of build scripts, then features don't compose.

Another name could be made specifically for build scripts or we can have a run-specific isolated directory like being discussed in that other issue.

**Comment by Ddystopia** [comment]
> `OUT_DIR` is created when build scripts are present for them to write into and others to read. Setting another value for proc-macros would be a breaking change. If we conditioned it on the absence of build scripts, then features don't compose.
> 
> Another name could be made specifically for build scripts or we can have a run-specific isolated directory like being discussed in that other issue.

Yeah, name is irrelevant, just something to store artifacts. But isn't the purpose of that tmpdir is to *not* be usable as `include`? And btw, just as with `OUT_DIR`, preserving contents (bets effort) between reruns is useful.

**Comment by epage** [comment]
Yes, if its for `include!` then `TMPDIR` wouldn't work.

In general for `include!`, we'd run into the issue of how do the crates that a proc-macro is being built-in know about stuff related to the definition of the proc-macro.  Another example is trying to support `$crate`.
